### PR TITLE
Reduce Alpaka event synchronization calls via EDMetadata

### LINKED ISSUE #902: Reduce Alpaka event synchronization calls via EDMetadata
See https://github.com/cms-sw/cmssw/issues/44769#issuecomment-2067257800

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

Prompted by https://github.com/cms-sw/cmssw/issues/44769#issuecomment-2067257800 this PR reduces the amount of blocking `alpaka::wait()` calls (leading to `cudaEventSynchronize()` on NVIDIA GPU) via the EDMetadata destructor. To check the impact I used the reproducer https://github.com/cms-sw/cmssw/issues/44769 (even if it eventually leads to an assertion failure). Initially the reproducer lead to 2470 calls to `cudaEventSynchronize()`.

The first commit avoids calling `alpaka::wait()` in `EDProducer::produce()` when no asynchronous work was launched (possible async work is identified by the use of `device::Event::queue()` in any way, including consuming a device side data product). In this case there is no need to synchronize the device and host. This commit reduced the number of `cudaEventSynchronize()` calls to 2284 (-7.5 %).

The second commit caches the alpaka::Event completion, so that `alpaka::wait()` does not need to be called again on an already-completed alpaka Event. This commit reduced the number of `cudaEventSynchronize()` calls to 2019 (-18 % wrt the starting point).

The third commit avoids calling `alpaka::wait()` on the destructor of an `EDMetadata` when the Queue of the `EDMetadata` object has been re-used by another `EDMetadata` (in an `EDProducer::produce()` that consumed the device-side data product holding the first `EDMetadata` object). Since the only purpose of the `alpaka::wait()` in the `~EDMetadata()` is to ensure all asynchronous work of an edm::Event has completed before the edm::Stream moves to the next edm::Event, for any Queue it is enough to `wait()` on the alpaka::Event that was last recorded on that Queue. This commit reduced the number of `cudaEventSynchronize()` calls to 1462 (-40 % wrt the starting point).

There is one case left (on purpose) where the `EDProducer::produce()` still calls `alpaka::wait()`. This happens when the `produce()` launches asynchronous work (i.e. calls `device::Event::queue()`), but does not produce any device-side data products. I can't imagine a good use case for such an EDProducer (except maybe a DQM-style module, but that would need a separate module base class in any case that could then deal with the necessary edm::Event-level synchronization 

I don't expect a huge impact from the reduction of these calls, as in the present cases in HLT the alpaka::Events should have completed by the time the `alpaka::wait()` is called. The stack traces in https://github.com/cms-sw/cmssw/issues/44769#issuecomment-2066613569 nevertheless show the `cudaEventSynchronize()` acquiring a lock even when the CUDA event was complete at the time the event was recorded, so at least this PR should reduce the contention on that lock.

Resolves https://github.com/cms-sw/framework-team/issues/902

#### PR validation:

`HeterogeneousCore/Alpaka{Core,Test}` unit tests pass. 

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Likely will be backported to 14_0_X

### CODE REVIEW

**Comment by makortel** [comment]
enable gpu

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
> Likely will be backported to 14_0_X

@fwyzard This PR might warrant some further testing on the HLT side. Let me know how you want to proceed (e.g. I have a 14_0_X-based branch already too).

File: HeterogeneousCore/AlpakaCore/README.md [hunk 1]
```diff
@@ -170,7 +170,7 @@ Also note that the `fillDescription()` function must have the same content for a
 * All Event data products in the host memory space are guaranteed to be accessible for all operations (after the data product has been obtained from the `edm::Event` or `device::Event`).
 * All EventSetup data products in the device memory space are guaranteed to be accessible only for operations enqueued in the `Queue` given by `device::Event::queue()` when accessed via the `device::EventSetup` (ED modules), or by `device::Record<TRecord>::queue()` when accessed via the `device::Record<TRecord>` (ESProducers).
 * The EDM Stream does not proceed to the next Event until after all asynchronous work of the current Event has finished.
-  * **Note**: currently this guarantee does not hold if the job has any EDModule that launches asynchronous work but does not explicitly synchronize or produce any device-side data products.
+  * **Note**: this implies if an EDProducer in its `produce()` function uses the `Event::queue()` or gets a device-side data product, and does not produce any device-side data products, the `produce()` call will be synchronous (i.e. will block the CPU thread until the asynchronous work finishes)
```
Comment (makortel) [inline HeterogeneousCore/AlpakaCore/README.md:173]: This update in the README actually holds (and is stronger) even without this PR, I just hadn't realized it.

Comment (fwyzard) [reply HeterogeneousCore/AlpakaCore/README.md:173]: My brain hurts, but I trust you on this :-)

**Comment by fwyzard** [comment]
The code changes look good.

I will try to test the impact on an HLT workflow.

**Comment by fwyzard** [comment]
here's the impact on the HLT (8 jobs, 32 threads / 24 streams each):

##### CMSSW_14_0_5_patch2
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   669.6 ±   0.2 ev/s (10000 events, 99.1% overlap)
   672.7 ±   0.2 ev/s (10000 events, 98.7% overlap)
   668.4 ±   0.3 ev/s (10000 events, 98.5% overlap)
   671.1 ±   0.2 ev/s (10000 events, 99.0% overlap)
 --------------------
   670.5 ±   1.9 ev/s
```

##### CMSSW_14_0_5_patch2 plus [cms-sw/cmssw#44841](https://github.com/cms-sw/cmssw/pull/44841)
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   671.1 ±   0.2 ev/s (10000 events, 99.4% overlap)
   671.9 ±   0.2 ev/s (10000 events, 98.7% overlap)
   673.5 ±   0.2 ev/s (10000 events, 98.8% overlap)
   675.6 ±   0.2 ev/s (10000 events, 98.2% overlap)
 --------------------
   673.0 ±   2.0 ev/s
```

Whether the small improvement is real or a fluctuation is hard to say - but anyway it goes in the right direction.

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by antoniovilela** [comment]
+1

**Pull request merged**